[GLUTEN-9106][VL] Add support for staticInvoke CharVarcharCodegenUtils. #9107

Yifeng-Wang opened the pull request:
## What changes were proposed in this pull request?

To add support for static function callls in Spark's CharVarcharCodegenUtils.java, we add expr matching pattern in Gluten ExpressionConverter.

(Fixes: \#9106)

## How was this patch tested?

Added test cases in Gluten's ScalarFunctionsValidateSuite.
Built Gluten Velox with Spark 3.4 (Spark3.4.4 in pom.xml default) and test sqls.

for test_varcharWriteSideCheck: 
```sql
create table srcvarchar(id string) stored as parquet;
create table tgt_vanilla(id varchar(3)) stored as parquet;
create table tgt(id varchar(3)) stored as parquet;

create table srcchar(id string) stored as parquet;
create table tgt_vanilla(id char(3)) stored as parquet;
create table tgt(id char(3)) stored as parquet;

insert into srcvarchar values ('a');
insert into srcvarchar values ('ab');
insert into srcvarchar values ('t  '); -- trail space but should retain
insert into srcvarchar values ('tra '); -- trail space but should trim
insert into srcvarchar values ('中  '); -- trailing space but should retain
insert into srcvarchar values ('中   '); -- trailing space but should trim to 3
insert into srcvarchar values ('中文中国'); -- will fail

insert into tgt select id from srcvarchar
```

for charWriteSideCheck: 
```sql
insert into srcchar values ('a');
insert into srcchar values ('ab');
insert into srcchar values ('快');
insert into srcchar values ('捷  ');
insert into tgt_vanilla select id from srcchar where id='abcd';
```

check plan to validate such operations offload to native and do not cause operator fallback:
![3668dcde999255823c0bfcfc9f55970](https://github.com/user-attachments/assets/9a5a37ce-8f6a-4dc3-bd63-a6f80845dcf0)

Yifeng-Wang commented:
@zzcclp @jinchengchenghh Hi, dear Zhichao and Jincheng, could you please review this PR? It addresses issue #9106 .
We are now creating a relative issue & PR at velox side, and we look forward to your feedbacks as well.
Thanks in advance, best regards,

jinchengchenghh commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -164,6 +164,32 @@ object ExpressionConverter extends SQLConfHelper with Logging {
                 child.map(replaceWithExpressionTransformer0(_, attributeSeq, expressionsMap)),
                 i)
           }
+        } else if (objectName.endsWith("CharVarcharCodegenUtils")) {
```
Please extract the logic for case i: StaticInvoke => to a new function, here will be much code

jinchengchenghh commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -109,6 +109,66 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+      expressionsMap: Map[Class[_], String]): ExpressionTransformer = {
+    val objectName = i.staticObject.getName.stripSuffix("$")
+
+    if (objectName.endsWith("UrlCodec")) {
```
Please use this format to simplify the code, and easy to implement others
```
val staticInvokeMap = Map("UrlCodec"-> Map("decode" -> ExpressionNames.URL_DECODE, "encode"-> ExpressionNames.URL_ENCODE ))
 val transformer = staticInvokeMap.filter(o => objectName.endsWith(o._1))
if (transformer.isEmpty) {
      throw
    }
```

jinchengchenghh replied:
I think most of the cases we need all the argument, UrlCodec is a special case, maybe because the backend only supports "UTF-8" , we should throw exception if 2nd argument is not "UTF-8"  but not.

So other common cases should use the `staticInvokeMap`, that can also help developer figure out the special case.

jinchengchenghh commented on shims/common/src/main/scala/org/apache/gluten/expression/ExpressionNames.scala:
```diff
@@ -356,4 +356,9 @@ object ExpressionNames {
   final val UDF_PLACEHOLDER = "udf_placeholder"
   final val UDAF_PLACEHOLDER = "udaf_placeholder"
+
+  // Spark Catalyst util functions
```
Spark StaticInvoke Catalyst util functions

jinchengchenghh commented:
Yes, you don't need to support it, just make sure it is fallback in CH backend.
The CH CI user and password is gluten/hN2xX3uQ4m, please resolve it.
Like the ExpressionMapping function `extraExpressionMappings` for each backend, we may need the  `extraStaticInvokeMappings`, I suppose the failed CH CI is cause by this, native validation does not fallback.

jinchengchenghh commented:
The password is in the CH backend document, but hard to find

jinchengchenghh commented:
Just add the new functions name to CH_BLACKLIST_SCALAR_FUNCTION like URL_DECODE

jinchengchenghh commented:
Yes, I think so. Add the substrait name check in replaceStaticInvokeWithExpressionTransformer@taiyang-li Do you have some suggestions that make CH backend fallback these unsupported functions

jinchengchenghh commented:
These function are not merged in velox, we cannot add a unit test to it, can we merge these function native implement first?

Yifeng-Wang commented:
**Summary**

##### **What has been done:**

**Ⅰ. Rebased previous proposal onto current main branch progress**

Rebased code onto current main branch. This PR still extractS logics for `case i: StaticInvoke` branch into a separated `replaceStaticInvokeWithExpressionTransformer`, but now an extra `validateandTransform` is introduced to do the backend validation. The logic remain the same which ensures a Transformer is instantiated only when the underlying backend supports this expression.



##### **Ⅱ.  Tests for the new round**

 - **Gluten Test Suite**

run `mvn clean install -Pbackends-velox -Pspark-3.5 -DwildcardSuites="org.apache.gluten.functions.ScalarFunctionsValidateSuite"`

(-Pspark-3.4 for Spark v3.4) . 

**Validation**:

Got BUILD SUCCESS and checked surefire test reports "TEST-org.apache.gluten.functions.ScalarFunctionsValidateSuite*" all pass.

![image.png](https://cdn.nlark.com/yuque/0/2025/png/25765883/1756086653863-b39366d9-379b-4956-a551-484d49f114ca.png?x-oss-process=image%2Fformat%2Cwebp)



- **Spark + Gluten + Velox tests** 

Test for all current supported StaticInvoke Expressions.

```SQL
-- url_decode
create table url_tbl(id string) stored as parquet;
insert into url_tbl values ('https%3A%2F%2Fspark.apache.org');
select url_decode(id) from url_tbl; -- https://spark.apache.org
-- url_encode
create table url_tbl2(id string) stored as parquet;
insert into url_tbl2 values ('https://spark.apache.org');
select url_encode(id) from url_tbl2;   -- https%3A%2F%2Fspark.apache.org

-- luhn_check, min Spark>=3.5
select luhn_check(id) from url_tbl;

-- base64 and unbase64
select unbase64(base64(id)) from url_tbl;

-- CharVarcharCodegenUtils functions (this PR)
create table srcvarchar(id string) stored as parquet;
insert into srcvarchar values ('ab');
create table tgtvarchar(id varchar(3)) stored as parquet;
create table tgtchar(id char(3)) stored as parquet;
-- varcharTypeWriteSideCheck
insert into tgtvarchar select id from srcvarchar;
-- charTypeWriteSideCheck
insert into tgtchar select id from srcvarchar;
-- readSidePadding
SELECT id FROM tgtchar;
```

- **Validation:** 

Verify Transformers within SQL execution plans without fallbacks and results correct.

Yifeng-Wang commented:
Hi @jinchengchenghh and @rui-mo!

I've just pushed the latest updates to this PR and have included a summary above to summarize the current status.

I understand this PR has been through a lengthy process with a pause in the middle, and I appreciate your patience and valuable feedback thus far. It's now ready for a second round of review.

When you have a moment, could you please take a look at the recent changes? Your feedback would be greatly appreciated.

Thanks! :)

Yifeng-Wang commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -144,6 +144,67 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+      GenericExpressionTransformer(exprName, childTransformers, i)
+    }
+
+    if (Seq("encode", "decode").contains(i.functionName) && i.objectName.endsWith("UrlCodec")) {
```
Yes, both the codes in origin main branch and this PR remain case-sensitive because they're comparing against actual Java method and class names.
![Uploading image.png…]()

Yifeng-Wang commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -144,6 +144,67 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+      )
+    }
+
+    if (Seq("encode", "decode").contains(i.functionName) && i.objectName.endsWith("Base64")) {
```
Yes, case-sensitive is intentional. Same as above.
<img width="1425" height="67" alt="image" src="https://github.com/user-attachments/assets/3e44797d-2f2a-41e7-9a79-f9b92587a601" />

Yifeng-Wang commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -144,6 +144,67 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+    }
+
+    if (
+      Set("varcharTypeWriteSideCheck", "charTypeWriteSideCheck", "readSidePadding").contains(
```
Yes, case-sensitive is intentional. Same as above.
<img width="1125" height="69" alt="image" src="https://github.com/user-attachments/assets/a45963f1-875d-47b2-aef2-07a5b0205546" />

Yifeng-Wang commented:
The latest commit added `excludeGlutenTest` to ClickHouseTestSettings.scala across different versions. The added expression support feature in this PR is only supported by Velox backend, while for ClickHouse backend, it fallbacks and thus should not match the expected err msg at gluten test suite.

rui-mo commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -144,6 +144,66 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+        val exprName = fn match {
+          case "varcharTypeWriteSideCheck" => ExpressionNames.VARCHAR_TYPE_WRITE_SIDE_CHECK
+          case "charTypeWriteSideCheck" => ExpressionNames.CHAR_TYPE_WRITE_SIDE_CHECK
+          case "readSidePadding" => ExpressionNames.READ_SIDE_PADDING
```
The mappings from Spark's name to Velox's name are more commonly to be put at https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitParser.cc#L387 to allow compatibility among different backends.

Yifeng-Wang replied:
@rui-mo Hi rui-mo, could you help give me some guidance on how I should update the codes?
I guess this is perhaps a static map to record method names (mismatching) ?

Yifeng-Wang replied:
To add some context, 
at Gluten side in `ExpressionNames.scala`. I had e.g.:
`final val VARCHAR_TYPE_WRITE_SIDE_CHECK = "varchar_type_write_side_check"` .
And at velox side, there is
```
registerFunction<
VarcharTypeWriteSideCheckFunction,
Varchar,
Varchar,
int32_t>
({prefix + "varchar_type_write_side_check"});
```
so I guess substrait function name (from `ExpressionNames.VARCHAR_TYPE_WRITE_SIDE_CHECK = "varchar_type_write_side_check"`) already matches the Velox registered function name (prefix + "varchar_type_write_side_check").

rui-mo commented on gluten-substrait/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:
```diff
@@ -144,6 +144,66 @@ object ExpressionConverter extends SQLConfHelper with Logging {
+        if (!BackendsApiManager.getValidatorApiInstance.doExprValidate(ExpressionNames.BASE64, i)) {
+          throw new GlutenNotSupportException(
+            s"Not supported to map current ${i.getClass} call on function: ${i.functionName}.")
+        }
```
Thanks, I understand.
